sagemaker-debugger/model_specific_realtime_analysis/bert_attention_head_view/bert_attention_head_view.ipynb

```diff
@@ -0,0 +1,538 @@
+    "        tensors[tname][step] = trial.tensor(tname).value(step, modes.EVAL)\n",
+    "num_heads = tensors[tname][step].shape[1]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next we get the query and key output tensor names:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "layers = []\n",
+    "layer_names = {}\n",
+    "\n",
+    "for index, (key, query) in enumerate(zip(trial.tensor_names(regex='.*key_output_'), trial.tensor_names(regex='.*query_output_'))):\n",
+    "    layers.append([key,query])\n",
+    "    layer_names[key.split('_')[1]] = index"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We also retrieve the string representation of the input tokens that were input into our model during validation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "input_tokens = trial.tensor('input_tokens').value(0, modes.EVAL)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Attention Head View\n",
+    "\n",
+    "The attention-head view shows the attention scores between different tokens. The thicker the line the higher the score. For demonstration purposes, we will limit the visualization to the first 20 tokens. We can select different attention heads and different layers. As training progresses attention scores change and we can check that by selecting a different step. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_tokens = 20\n",
+    "view = attention_head_view.AttentionHeadView(input_tokens, \n",
+    "                                             tensors,  \n",
+    "                                             step=trial.steps(modes.EVAL)[0], \n",
+    "                                             n_tokens=n_tokens)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "interactive(view.select_layer, layer=tensor_names)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "interactive(view.select_head, head=np.arange(num_heads))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "interactive(view.select_step, step=trial.steps(modes.EVAL))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following code cell updates the dictionary `tensors`  with the latest tensors from training the job. Once the dict is updated we can go to above code cell `attention_head_view.AttentionHeadView` and re-execute this and subsequent cells in order to plot latest attentions."
```

Comment:
Another point here: maybe it makes sense to restructure this code so that users won't have to jump between cells in order to update tensors -> plot -> update tensors -> plot and so forth? I think there are some examples of how it could be achieved in some other notebooks of real time analysis. Not a blocker though.

Reply:
I originally planned to do that, but having multiple IPython widgets in the same notebook cell does not seem to work. So I seperated them. Not sure if there is a workaround?

Reply:
I am currently testing the notebook in JupyterLab. While the notebook is working fine in Jupyter, it seems widgets are causing problems in JupyterLab. I would suggest to wait with merging of this PR.

Reply:
I figured out how to fix the issues in JupyterLab and added instructions to the notebook.